# GameGuide — Data Understanding

## Step 1: Dataset Loading and Inspection

The objective of this step is to understand the structure, quality,
data types, missing values, and unusual formats present in the raw
Steam Games dataset before performing any data cleaning.

We will inspect:

- Dataset dimensions
- Column names
- Data types
- Sample records
- Missing values
- List-like columns
- Price and currency information
- Duplicate games
- Potential data quality problems

In [1]:
import numpy as np
import pandas as pd

In [2]:
DATA_PATH = "../data/raw/games_march2025_full.csv"
df = pd.read_csv(DATA_PATH, low_memory=False)
print("Data Loaded Successfully!")

Data Loaded Successfully!


In [3]:
df.shape

(94948, 47)

In [4]:
df.head()

,appid,name,release_date,required_age,price,dlc_count,detailed_description,about_the_game,short_description,reviews,...,average_playtime_2weeks,median_playtime_forever,median_playtime_2weeks,discount,peak_ccu,tags,pct_pos_total,num_reviews_total,pct_pos_recent,num_reviews_recent
0,730,Counter-Strike 2,2012-08-21,0,0.00,1,"For over two decades, Counter-Strike has offer...","For over two decades, Counter-Strike has offer...","For over two decades, Counter-Strike has offer...",NaN,...,879,5174,350,0,1212356,"{'FPS': 90857, 'Shooter': 65397, 'Multiplayer'...",86,8632939,82,96473
1,578080,PUBG: BATTLEGROUNDS,2017-12-21,0,0.00,0,"LAND, LOOT, SURVIVE! Play PUBG: BATTLEGROUNDS ...","LAND, LOOT, SURVIVE! Play PUBG: BATTLEGROUNDS ...",Play PUBG: BATTLEGROUNDS for free. Land on str...,NaN,...,0,0,0,0,616738,"{'Survival': 14838, 'Shooter': 12727, 'Battle ...",59,2513842,68,16720
2,570,Dota 2,2013-07-09,0,0.00,2,"The most-played game on Steam. Every day, mill...","The most-played game on Steam. Every day, mill...","Every day, millions of players worldwide enter...",“A modern multiplayer masterpiece.” 9.5/10 – D...,...,1536,898,892,0,555977,"{'Free to Play': 59933, 'MOBA': 20158, 'Multip...",81,2452595,80,29366
3,271590,Grand Theft Auto V Legacy,2015-04-13,17,0.00,0,"When a young street hustler, a retired bank ro...","When a young street hustler, a retired bank ro...",Grand Theft Auto V for PC offers players the o...,NaN,...,771,7101,74,0,117698,"{'Open World': 32644, 'Action': 23539, 'Multip...",87,1803832,92,17517
4,488824,Tom Clancy's Rainbow Six® Siege,2015-12-01,17,19.99,9,Edition Comparison Ultimate Edition The Tom Cl...,“One of the best first-person shooters ever ma...,"Tom Clancy's Rainbow Six® Siege is an elite, t...",NaN,...,0,0,0,0,0,"{'FPS': 8082, 'Multiplayer': 6139, 'Tactical':...",84,1168404,76,13017


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94948 entries, 0 to 94947
Data columns (total 47 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   appid                     94948 non-null  int64  
 1   name                      94946 non-null  object 
 2   release_date              94948 non-null  object 
 3   required_age              94948 non-null  int64  
 4   price                     94948 non-null  float64
 5   dlc_count                 94948 non-null  int64  
 6   detailed_description      89522 non-null  object 
 7   about_the_game            89499 non-null  object 
 8   short_description         89599 non-null  object 
 9   reviews                   10428 non-null  object 
 10  header_image              94948 non-null  object 
 11  website                   41194 non-null  object 
 12  support_url               44185 non-null  object 
 13  support_email             78848 non-null  object 
 14  window

In [6]:
df.columns.tolist()

['appid',
 'name',
 'release_date',
 'required_age',
 'price',
 'dlc_count',
 'detailed_description',
 'about_the_game',
 'short_description',
 'reviews',
 'header_image',
 'website',
 'support_url',
 'support_email',
 'windows',
 'mac',
 'linux',
 'metacritic_score',
 'metacritic_url',
 'achievements',
 'recommendations',
 'notes',
 'supported_languages',
 'full_audio_languages',
 'packages',
 'developers',
 'publishers',
 'categories',
 'genres',
 'screenshots',
 'movies',
 'user_score',
 'score_rank',
 'positive',
 'negative',
 'estimated_owners',
 'average_playtime_forever',
 'average_playtime_2weeks',
 'median_playtime_forever',
 'median_playtime_2weeks',
 'discount',
 'peak_ccu',
 'tags',
 'pct_pos_total',
 'num_reviews_total',
 'pct_pos_recent',
 'num_reviews_recent']

In [7]:
columns_info = pd.DataFrame({
    "columns" : df.columns,
    "dtype" : df.dtypes,
    "non_null" : df.notna().sum().values,
    "missing" : df.isna().sum().values
})

columns_info

,columns,dtype,non_null,missing
appid,appid,int64,94948,0
name,name,object,94946,2
release_date,release_date,object,94948,0
required_age,required_age,int64,94948,0
price,price,float64,94948,0
dlc_count,dlc_count,int64,94948,0
detailed_description,detailed_description,object,89522,5426
about_the_game,about_the_game,object,89499,5449
short_description,short_description,object,89599,5349
reviews,reviews,object,10428,84520


In [8]:
missing_percentage = (
    df.isna().mean() * 100
).round(2).sort_values(ascending= False)

missing_percentage.head(20)

score_rank              99.96
metacritic_url          96.23
reviews                 89.02
notes                   82.46
website                 56.61
support_url             53.46
support_email           16.96
about_the_game           5.74
detailed_description     5.71
short_description        5.63
dlc_count                0.00
appid                    0.00
price                    0.00
required_age             0.00
release_date             0.00
windows                  0.00
header_image             0.00
name                     0.00
linux                    0.00
metacritic_score         0.00
dtype: float64

In [19]:
missing_table = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_percentage": (df.isna().mean() * 100).round(2)
}).sort_values("missing_percentage", ascending=False)

missing_table

,missing_count,missing_percentage
score_rank,94909,99.96
metacritic_url,91372,96.23
reviews,84520,89.02
notes,78296,82.46
website,53754,56.61
support_url,50763,53.46
support_email,16100,16.96
about_the_game,5449,5.74
detailed_description,5426,5.71
short_description,5349,5.63


In [20]:
tricky_columns = [
    "genres",
    "categories",
    "tags",
    "discount",
    "estimated_owners",
    "release_date",
    "price",
    "supported_languages"
]

for col in tricky_columns:
    print("=" * 70)
    print(f"COLUMN: {col}")
    print(df[col].head(5).to_list())

COLUMN: genres
["['Action', 'Free To Play']", "['Action', 'Adventure', 'Massively Multiplayer', 'Free To Play']", "['Action', 'Strategy', 'Free To Play']", "['Action', 'Adventure']", "['Action']"]
COLUMN: categories
["['Multi-player', 'Cross-Platform Multiplayer', 'Steam Trading Cards', 'Steam Workshop', 'In-App Purchases', 'Valve Anti-Cheat enabled', 'Stats', 'Remote Play on Phone', 'Remote Play on Tablet', 'Remote Play on TV', 'Steam Timeline']", "['Multi-player', 'PvP', 'Online PvP', 'Stats', 'Remote Play on Phone', 'Remote Play on Tablet']", "['Multi-player', 'Co-op', 'Steam Trading Cards', 'Steam Workshop', 'SteamVR Collectibles', 'In-App Purchases', 'Valve Anti-Cheat enabled', 'Steam Timeline']", "['Single-player', 'Multi-player', 'PvP', 'Online PvP', 'Co-op', 'Online Co-op', 'Steam Achievements', 'Full controller support', 'Remote Play on Phone', 'Remote Play on Tablet', 'Remote Play on TV']", "['Single-player', 'Multi-player', 'PvP', 'Online PvP', 'Co-op', 'Online Co-op', 'Full

In [28]:
format_summary = {
    "genres": "String representation of list",
    "categories": "String representation of list",
    "tags": "String representation of dictionary: tag -> vote count",
    "discount": "Numeric percentage/integer",
    "estimated_owners": "String range: minimum - maximum",
    "release_date": "String date: YYYY-MM-DD",
    "price": "Numeric decimal price",
    "supported_languages": "String representation of list"
}

pd.Series(format_summary, name="Observed format")

genres                                     String representation of list
categories                                 String representation of list
tags                   String representation of dictionary: tag -> vo...
discount                                      Numeric percentage/integer
estimated_owners                         String range: minimum - maximum
release_date                                     String date: YYYY-MM-DD
price                                              Numeric decimal price
supported_languages                        String representation of list
Name: Observed format, dtype: object

In [14]:
df[df["name"].str.contains(
    "Elden Ring",
    case=False,
    na=False
)][
    ["appid", "name", "price", "release_date",
     "genres", "developers", "publishers"]
].head()

,appid,name,price,release_date,genres,developers,publishers
16,1245620,ELDEN RING,59.99,2022-02-24,"['Action', 'RPG']","['FromSoftware, Inc.']","['FromSoftware, Inc.', 'Bandai Namco Entertain..."


In [21]:
print(df["price"].describe())

count    94948.000000
mean         6.911444
std         13.071148
min          0.000000
25%          0.990000
50%          3.990000
75%          9.990000
max        999.980000
Name: price, dtype: float64


In [22]:
print(df["price"].value_counts().head(20))

price
0.00     19420
4.99      9714
0.99      8381
9.99      8114
1.99      6292
2.99      6256
3.99      4484
14.99     3786
19.99     3619
5.99      3140
6.99      2859
7.99      2627
8.99      1432
11.99     1335
24.99     1204
29.99     1180
12.99     1110
0.49       631
10.99      590
39.99      545
Name: count, dtype: int64


In [23]:
print(
    df.loc[
        df["name"].str.contains(
            "Elden Ring|Portal 2|Counter-Strike 2|Grand Theft Auto V",
            case=False,
            na=False
        ),
        ["name", "price"]
    ].head(20)
)

                             name  price
0                Counter-Strike 2   0.00
3       Grand Theft Auto V Legacy   0.00
16                     ELDEN RING  59.99
1117  Grand Theft Auto V Enhanced  14.99
8210   Portal 2 - The Final Hours   1.99


In [17]:
appid_duplicates = df["appid"].duplicated().sum()
name_duplicates = df["name"].duplicated().sum()

print("Duplicate appids:", appid_duplicates)
print("Duplicate names:", name_duplicates)

Duplicate appids: 0
Duplicate names: 756


In [18]:
print("Missing appid:", df["appid"].isna().sum())
print("Missing name:", df["name"].isna().sum())

Missing appid: 0
Missing name: 2


In [24]:
df[df["name"].isna()][["appid", "name", "release_date", "price"]]

,appid,name,release_date,price
26053,396420,NaN,2016-11-01,0.00
36037,1116910,NaN,2019-09-25,6.99


In [25]:
duplicate_names = (
    df[df["name"].duplicated(keep=False)]
    .sort_values("name")
)

duplicate_names[["appid", "name", "price"]].head(30)

,appid,name,price
9444,842420,A Walk in the Woods,0.00
21722,557900,A Walk in the Woods,0.00
86485,1768480,ASTRA,0.99
44434,2691830,ASTRA,2.99
78357,1522190,ASTRA,0.00
94792,2786120,Abyss,2.99
59543,1633770,Abyss,6.99
34427,2493960,Abyss,24.95
3854,743640,Achievement Clicker,0.99
26690,1229730,Achievement Clicker,0.99


In [26]:
df["name"].value_counts().head(20)

name
Shadow of the Tomb Raider: Definitive Edition    20
Alone                                             6
Aurora                                            5
Echoes                                            5
Tom Clancy's Rainbow Six® Siege                   5
EA SPORTS FC™ 24                                  4
Delirium                                          4
Bounce                                            4
Exodus                                            4
The House                                         4
Arena                                             4
No Way Out                                        4
Hide and Seek                                     4
Escape                                            4
Loop                                              4
EA SPORTS FC™ 25                                  4
The Unknown                                       4
The Castle                                        3
Memento Mori                                      3
Requiem

In [27]:
df["reviews"].dropna().head(10).tolist()

['“A modern multiplayer masterpiece.” 9.5/10 – Destructoid “Once you start to learn its secrets, there’s a wild and exciting variety of play here that’s unmatched, even by its peers.” 9.4/10 – IGN “Dota 2 is possibly the only competitive free-to-play game that is totally uncompromised by its business model.” 90/100 – PC Gamer ',
 "“Rust is one of the cruelest games on Steam, and that's what makes it so compelling.” PC Gamer “Usually, people will just kill you on sight, but not always.” Rock Paper Shotgun “For competitive players who want a mix of survival and crafting, Rust is the best there is” IGN ",
 "“It brings back the silly glee of unbridled experimentation, intuitive learning and playful social interactions that most of us haven't experienced since childhood. For that, it's essential.” 88 – PC Gamer ",
 "“The champion of Battle Royales.” 9/10 – GameSpot “The best Battle Royale game.” 5/5 – GamesRadar+ “Free, friendly, and consistently fun. It's the best Battle Royale game availa

# Problems Found in the Raw Dataset

Based on the initial inspection, the following data-quality issues
were identified:

1. The dataset contains 94,948 records and 47 columns, which differs
   from the expected 89,625 records and 46 columns in the project plan.

2. Two records have missing game names.

3. There are 756 duplicate game names, while appid has no duplicates.

4. Several columns have substantial missing values:
   score_rank, metacritic_url, reviews, notes, website and support_url.

5. The description fields have moderate missing values:
   short_description, about_the_game and detailed_description.

6. genres and categories are stored as string representations of lists.

7. tags is stored as a string representation of a dictionary containing
   tag vote counts.

8. estimated_owners is stored as a string range rather than a numeric value.

9. release_date is stored as an object/string and will need conversion
   to datetime.

10. Price values are decimal monetary values and need to be interpreted
    consistently before INR conversion.

11. Some numeric fields may contain placeholder values such as 0 that
    need to be investigated rather than treated automatically as valid
    measurements.